access/email/app.go

```diff
@@ -0,0 +1,516 @@
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if !ok {
+		logger.Get(ctx).Debug("Failed to post reply: plugin data is missing")
+		return nil
+	}
+	reviews := reqReviews[oldCount:]
+	if len(reviews) == 0 {
+		return nil
+	}
+
+	errors := make([]error, 0, len(reviews))
+	for _, review := range reviews {
+		threadsSent, err := a.client.SendReview(ctx, threads, reqID, reqData, review)
+		if err != nil {
+			errors = append(errors, err)
+		}
+		logger.Get(ctx).Infof("New review for request %v by %v is %v", reqID, review.Author, review.ProposedState.String())
+		logSentThreads(ctx, threadsSent, "new review")
+	}
+
+	return trace.NewAggregate(errors...)
+}
+
+// sendResolution updates the messages status and sends message when request has been resolved
+func (a *App) sendResolution(ctx context.Context, reqID string, resolution Resolution) error {
+	log := logger.Get(ctx)
+
+	var pluginData PluginData
+	ok, err := a.modifyPluginData(ctx, reqID, func(existing *PluginData) (PluginData, bool) {
+		// If plugin data is empty or missing email message timestamps, we cannot do anything.
+		if existing == nil {
+			return PluginData{}, false
+		}
+		if pluginData = *existing; len(pluginData.EmailThreads) == 0 {
+			return PluginData{}, false
+		}
+
+		// If resolution field is not empty then we already resolved the incident before. In this case we just quit.
+		if pluginData.RequestData.Resolution.Tag != Unresolved {
+			return PluginData{}, false
+		}
+
+		// Mark plugin data as resolved.
+		pluginData.Resolution = resolution
+		return pluginData, true
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if !ok {
+		log.Debug("Failed to update messages: plugin data is missing")
+		return nil
+	}
+
+	reqData, threads := pluginData.RequestData, pluginData.EmailThreads
+
+	threadsSent, err := a.client.SendResolution(ctx, threads, reqID, reqData)
+	logSentThreads(ctx, threadsSent, "request resolved")
+
+	log.Infof("Marked request as %s and sent emails!", resolution.Tag)
+
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// modifyPluginData performs a compare-and-swap update of access request's plugin data.
+//
+// Callback function parameter is nil if plugin data hasn't been created yet.
+//
+// Otherwise, callback function parameter is a pointer to current plugin data contents.
+// Callback function return value is an updated plugin data contents plus the boolean flag
+// indicating whether it should be written or not.
+//
+// Note that callback function fn might be called more than once due to retry mechanism baked in
+// so make sure that the function is "pure" i.e. it doesn't interact with the outside world:
+// it doesn't perform any sort of I/O operations so even things like Go channels must be avoided.
+//
+// Indeed, this limitation is not that ultimate at least if you know what you're doing.
+func (a *App) modifyPluginData(ctx context.Context, reqID string, fn func(data *PluginData) (PluginData, bool)) (bool, error) {
```

Comment:
Not necessary for _this_ PR, but a version of this seems to exist for each plugin. Is there a way to factor it out into a common package - or are there subtle differences between the plugin versions that I'm not seeing?